ks: main action for items

alex40724 opened the pull request:
This adds a main action to KS items to provide a button for the most important action of an item.

This is needed to implement the feature https://docu.ilias.de/goto_docu_wiki_wpage_7444_1357.html

The tests are currently failing. I will fix them and add new ones as soon as we decide on the interface and the rendering.

I am sure there will be a more appropriate sass variable for defining the margin, currently it says
`margin-right: $il-padding-small-horizontal;` Any suggestions?

<img width="717" alt="Bildschirmfoto 2023-08-31 um 12 04 12" src="https://github.com/ILIAS-eLearning/ILIAS/assets/1221841/d7e4cd4b-5da6-4ce5-90bf-cf6ddc520aa9">

alex40724 committed:
fixed bug #32128: Accordeon -vertical does not react after opening one panel

lzehnder committed:
[FEATURE] ILCORE-706: Improve File Information on Info Page

lzehnder committed:
[FEATURE] ILCORE-712: Important Information for File Object

kergomard committed:
Object: Move Tiles to IRSS

lzehnder committed:
[FIX] Modules/File: fixed download launcher not being shown if there is no important information

lzehnder committed:
[IMPROVEMENT] Modules/File: improved db update step for important file info

chfsx committed:
[FIX] prevent some alternative php extensions

lzehnder committed:
[FIX] Modules/File: fixed issue where stripping of extension would remove too much from a title containig a full stop

lzehnder committed:
Revert "[FIX] Modules/File: fixed issue where stripping of extension would remove too much from a title containig a full stop"

This reverts commit 42bd6f5c750014e1bc547cb40984be2fac1bc30c.

lzehnder committed:
[FIX] Modules/File: Fixed issues with stripping of file extension from title

nhaagen committed:
TA: use PresentationTable for results

nhaagen committed:
TA: use PresentationTable for results, question icon

nhaagen committed:
TA: use PresentationTable for results, add ContentStyle

nhaagen committed:
TA: use PresentationTable for results, rebase

nhaagen committed:
TA: use PresentationTable for results, remove comments

nhaagen committed:
PRG: 37678, sort before offset/limit on entire memberlist

nhaagen committed:
UI, Docu: allow docu-entries for array-returns and outside of UI/Component

lzehnder committed:
WIP: started implementation of kiosk mode for file

alex40724 committed:
ks: main action for items

catenglaender committed:
9/TA/ResultPresentation listing style

catenglaender commented on src/UI/templates/default/Item/tpl.item_standard.html:
```diff
@@ -36,7 +36,10 @@
 		<div class="media-body">
 			<!-- END lead_start_icon -->
 			<div class="il-item-title">{TITLE}</div>
+			<div class="il-item-actions" style="display: inline-block; float: right;">
```
This should work without the inline styles.

Alternatively, I would suggest using classes from the brand new layout tool "element-bar" (located in `templates/default/050-layout/_layout_element-bar.scss`), which is designed for exactly this kind of button bars. In this case you can remove all CSS but `float: right` from `.il-item-actions` in `_ui-component_item.scss`. Just assigning the classes should suffice to place them next to each other with the margin that other button bars have (or will have) as well:
```html
<div class="il-item-actions l-bar__container">
    <div class="l-bar__element">{MAIN_ACTION}</div>
    <div class="l-bar__element">{ACTIONS}</div>
</div>
```

catenglaender commented on templates/default/070-components/UI-framework/Item/_ui-component_item.scss:
```diff
@@ -43,6 +43,14 @@ $il-item-gap-xs: 1px;
 	}
 
+	.il-item-actions {
+		display: inline-block;
```
As far as I know `float: right` forces `display: inline-block` to be rendered as `display: block` anyway, so I suspect that we can do without this line.

If we decide to use the element-bar layout tool for spacing the buttons, almost no CSS code is needed on `.il-item-actions`.

catenglaender commented on src/UI/templates/default/Item/tpl.item_standard.html:
```diff
@@ -36,7 +36,10 @@
 		<div class="media-body">
 			<!-- END lead_start_icon -->
 			<div class="il-item-title">{TITLE}</div>
```
If the title is supposed to float around the actions, the title div has to come after the actions. This has been in the wrong order before this PR and is already broken in ILIAS 8.

catenglaender commented on templates/default/070-components/UI-framework/Item/_ui-component_item.scss:
```diff
@@ -43,6 +43,14 @@ $il-item-gap-xs: 1px;
+		display: inline-block;
+		float: right;
+		.btn-default {
+			margin-right: $il-padding-small-horizontal;
```
This would also push the last button in the row to the left not aligning with the edge of the padding.

I would suggest either adding
```scss
&:last-child {
    margin-right: 0;
}

```
or using the element-bar as it handles the end of the line (and as a bonus also adds bottom margin for top rows should the bar ever stack over multiple lines... which is unlikely with just two buttons, but I like to cover these multi-line edge cases).

catenglaender reviewed (commented):
Awesome! @yvseiler and I were thinking about main action buttons on items and entities for a while now. That's a useful feature for prioritizing an action. Thank you very much.

Here you can find a few code comments from my html/css focused perspective. I am a bit nitpicky here considering that one comment is also talking about a flaw in the current implementation.

There is a brand new element-bar scss layout tool that could be quite useful for controlling the button spacing from a central pattern. For two buttons it's not really needed, but it might remove the need to add any scss code with this PR.

I am looking forward to having main action buttons like this in the item and other UI components, as I don't think the often most important action should be burried in a dropdown.

chfsx committed:
[FIX] removed some unnecessary code

chfsx committed:
[FIX] some headers in files and code style

chfsx committed:
General PHP 8.1 refactoring

chfsx committed:
[FIX] removed unused method

catenglaender committed:
9/TA/ResultPresentation tweaked padding

Amstutz commented:
Hi @alex40724

Thx a lot for this PR. As @catenglaender we agree that this is a very useful feature. 

Please or indicate why not:

- [x] Inline Styles: Please removed them, as suggested by @catenglaender in the comment.
- [x] Layout tool: Use the element-bar layout tool for spacing the buttons, as suggested by @catenglaender in the comment.
- [x] Margin right: Implement one of the suggested option by @catenglaender.

Please change

- [x] Tests: If accepted by JF, add the tests
- [x] Testrail: If accepted by JF, please add an according testrail case to update the Testcase for the Standard Item.

Thx
@amstutz, @thibsy, @klees

nhaagen committed:
TA: fix TestObj Mock

nhaagen committed:
TA: some construction cleanup

nhaagen committed:
TA: use PresentationTable for results, user view

nhaagen committed:
TA: Result Toolbar with print-button

nhaagen committed:
TA: dedicated Feedbackoptions for inline/general (MC, SC)

nhaagen committed:
TA: relay text-only option to result presentation

nhaagen committed:
TA: use PresentationTable for results, add content for recapitulation

nhaagen committed:
TA: use PresentationTable for results, icons OK/MostlyOK/NotOK

nhaagen committed:
TA: use PresentationTable for results, translate labels

nhaagen committed:
TA: use PresentationTable for results, wrap answers in std. question container

nhaagen committed:
TA: use PresentationTable for results, show inline feedback dependent on setting

nhaagen committed:
TA: flag questions (instead of marking them)

catenglaender committed:
9/TA/ResultPresentation show listing in one row

catenglaender committed:
9/TA/ResultPresentation viewcontrol design

matthiaskunkel commented:
Jour Fixe, 18 SEP 2023 : We highly appreciate this suggestion and accept the PR for trunk. Please allow also Standard Links as action. Alexander will try to consider all comments made by Ferdinand and confirmed that he will add tests and a testcase as required by Timon.

nhaagen committed:
TA: integrate questionservice (after merge)

lukastocker committed:
LSO: 36631 add error message if Download Multiple Objects is used. (#36631)

lzehnder committed:
[FEATURE] ILCORE-696: Target Language for Administrative Notifications

klees committed:
Docs: use npm clean-install instead of install

thibsy committed:
[FIX] CI: install npm packages using clean-install.

lukastocker committed:
UI Tests:  adjust function names where snake case is still used.

klees commented:
@alex40724 Just want to make sure that we have no deadlock here. IMO the potato is in your hand, right?

lzehnder committed:
[FIX]: made some improvements regarding the implementation of ADN target languages

thibsy committed:
[FEATURE] UI: add chunked-upload for file-inputs.

chfsx committed:
Implemented WOPI

chfsx committed:
Implemented StaticURL

chfsx committed:
Implemented new RBAC operation

lzehnder committed:
[FEATURE] Centralizing online/offline status: Bibliographic

lzehnder committed:
[FEATURE] Centralizing online/offline status: File

# Conflicts:
#	Modules/File/classes/class.ilObjFileGUI.php

chfsx committed:
Implemented Draft Revisions

chfsx committed:
Public Interface for Draft Revisions

chfsx committed:
Use Draft Revisions in File Object

chfsx committed:
Edit Action in Dropdown

thibsy committed:
[FEATURE] FileServices: add global upload policies for customizable max file-sizes.

chfsx committed:
[FIX] sorting of revisions

chfsx committed:
[FIX] language variable

chfsx committed:
[FIX] get correct revision

nhaagen committed:
TA: use PresentationTable for results, rebase and fix

nhaagen committed:
TA: PresentationTable, reset params for getSolutionOutput in loop

nhaagen committed:
TA: use PresentationTable for results, fix icons

kergomard committed:
UI: Fix Race Condition When Loading Image

chfsx committed:
Improved JS

chfsx committed:
Support onlyoffice URLs

chfsx committed:
Save on Close

chfsx committed:
[FIX] session ttl

nhaagen committed:
TA, Result Settings: remove input groups and singlepage options

nhaagen committed:
UI, PresentationTable: fix tests

nhaagen committed:
UI, PresentationTable: remove ViewControls for print

nhaagen committed:
TA: use PresentationTable for results, remove comments

lzehnder committed:
[FIX] Reintroduced method getNamedObjectives to prevent error

tfamula committed:
Delete test skill data when deleting competences

chfsx committed:
Fixed Update Step Number

chfsx committed:
[FIX] missing LP toggle on infoscreen

fhelfer committed:
fix ilias undefined filedatamail checkreadwrite

Uwe-Kohnle committed:
modified scorm2004 buildrte

Uwe-Kohnle committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

kergomard committed:
Test: Duplicate Update Step

This is necessary, because there was a little mixup with
the update steps. It does not hurt, but will mostly be
unnecessary.

kergomard committed:
User: Fix Settings Should be Initialized

See: https://mantis.ilias.de/view.php?id=38233

kergomard committed:
User: Clean-Up ilUserXMLWriter

tfamula committed:
Delete test skill data when deleting competences

dsstrassner committed:
Change Label, as in the example only a date will be picked (#6429)

tfamula committed:
Fix copyright

nhaagen committed:
TA: PresentationTable for results, separate Result-/PresentationFactory

Uwe-Kohnle committed:
small changes to LTIConsumer with fixer

Uwe-Kohnle committed:
small changes for CmiXapi with fixer

tfamula committed:
Adjust skill readme

tfamula committed:
Merge branch 'trunk' of github.com:ILIAS-eLearning/ILIAS into trunk

nhaagen committed:
TA: PresentationTable for results, some renaming and use-statements

smeyer-ilias committed:
search: maven support;

smeyer-ilias committed:
Merge branch 'trunk' of github.com:ILIAS-eLearning/ILIAS into trunk

Uwe-Kohnle committed:
updates for Services/LTI with fixer

smeyer-ilias committed:
search: removed deprecated libs, files

nhaagen committed:
TA: PresentationTable, remove mutators from PassResultsSettings

kergomard committed:
File: Move Object Specific Tiles to Properties

nhaagen committed:
TA: #38270, declare class properties

kergomard committed:
Test: Clean-Up in ilTestNavigationToolbarGUI

kergomard committed:
Test: Fix Check Answer Button

kergomard committed:
Test: Fix Variable Instead of Class-Name

kergomard committed:
Test: Fix Undelcared Class-Property

nhaagen committed:
User: Increase length of firstname/lastname/email

See: https://mantis.ilias.de/view.php?id=37887

thibsy committed:
[FIX] Services\Form: removed import statements from Form.js.

matthiaskunkel committed:
Added final version of copyright icon

alex40724 committed:
portfolio/assignments: separate with/out template creation; Add link to submitted portfolio in notification mail; Change Shared status after submitting; Add Mailto Feature in list of Portfolio of Other Users; Confirmation Mail to Sharer after Sharing a Portfolio; notify graded members; Add Open Submission option in Participant View

Uwe-Kohnle committed:
updated for PHP and UI

smeyer-ilias committed:
search: libs updates; removed some deprecations

smeyer-ilias committed:
Merge branch 'trunk' of github.com:ILIAS-eLearning/ILIAS into trunk

thibsy committed:
[FIX] UI: fix chunk-sizes and filesize calculation.

thibsy committed:
[FEATURE] File: add chunked-upload support.

mbecker-databay committed:
Show Only Points

mbecker-databay committed:
Show Only Points

katringross committed:
fixed mantis bugs 38115, 38116 (#6432)

* fixed plugin lang vars
* fixed copyright

tfamula committed:
Fix property name

tfamula committed:
Merge branch 'trunk' of github.com:ILIAS-eLearning/ILIAS into trunk

gvollbach committed:
9 remove pdf service  (#6438)

* Remove pdf classes
* Remove lang vars
* Fix readme
* Remove tcpdf lib
* Fixed composer

matheuszych committed:
[FEATURE] UI: add customisation of submit- and cancel-labels. (#6355)

* Added methods to Launcher Interface
* Added launcher modal buttons labels
* Added Unit-Tests

---------

Co-authored-by: Matheus Zych <mzych@databay.de>

chfsx committed:
added missing content of file-preview readme

fhelfer committed:
remove ass suggestion type textCC

fhelfer committed:
remove ass suggestions records of type text from database

fhelfer committed:
remove suggestion type text tests

fhelfer committed:
copyright

chfsx committed:
Feature/9/dcl date fomulas (#6419)

* Implemented Date Fields in Formula Fields

* ilException not found

* [FIX] MIN/MAX in date formulas

* [FIX] added comment on why we override this method

* [FIX] removed not needed null check

* [FIX] removed disabled code

* [FIX] added byline addition to explain how to handle dates and integers

* Added ROADMAP

* [FIX] string tokens

chfsx committed:
[FIX] data-urls for in-memory-streams

kergomard committed:
Test: Remove Month From Waiting Time

chfsx committed:
[FIX] switched button labels

chfsx committed:
[FIX] missing file infos

chfsx committed:
[FIX] 0038226: fixed renaming of files in exercise

chfsx committed:
[FIX] 0037692: lang var 'mm_rep_tree_view' vs. 'mm_repo_tree_view'

mbecker-databay committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

catenglaender committed:
9/UI/buttons no press transform on active state (#6357)

Co-authored-by: Timon Amstutz <timon.amstutz@gmail.com>

Amstutz committed:
Scss Recompile

catenglaender committed:
9/UI/symbol/icon width on mobile 38166 (#6395)

Co-authored-by: Timon Amstutz <timon.amstutz@gmail.com>

alex40724 committed:
removed unnecessary includes

catenglaender committed:
9/UI/image responsive max width fix 37743 (#6397)

catenglaender committed:
9/UI/item drodown pull 38054 (#6400)

* 9/UI/SCSS tool modern clearfix

* 9/UI/SCSS float and clear in hacks

* 9/UI/Item Dropdown pull all the way to right 38054

* 9/SCSS compile CSS

catenglaender committed:
9/UI/keyboard focus outline tweaks 37635 (#6430)

Co-authored-by: Timon Amstutz <timon.amstutz@gmail.com>

Amstutz committed:
Scss Recompile

kergomard committed:
User: Correct Function Parameter Type for Disabled

See: https://mantis.ilias.de/view.php?id=38283

thibsy committed:
[FIX] UI: remove obsolete unit test again.

alex40724 committed:
portfolio: fixed buttons

thibsy committed:
[FEATURE] UI: added chunked-upload support for file-inputs.

Amstutz committed:
System Styles, fix of #36856

chfsx committed:
Merge branch 'feature/9/kiosk-mode-for-file' into trunk

chfsx committed:
[FIX] login parameter handling

chfsx committed:
Added Readme and Roadmap

chfsx committed:
[FIX] ResourceBuilder call

chfsx committed:
[FIX] db update

chfsx committed:
[FIX] merge issues

chfsx committed:
Prepared Callback in JS

Uwe-Kohnle committed:
delete not needed templates and remove warnings

kergomard committed:
Test: Add Time to DateTime and Duration Inputs

See: https://mantis.ilias.de/view.php?id=38265

kergomard committed:
UI: Fix datetime-local Value in Wrong Format

See:
* https://mantis.ilias.de/view.php?id=38265
* https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date
* https://developer.mozilla.org/en-US/docs/Web/HTML/Date_and_time_formats

Uwe-Kohnle committed:
add ui-elements and remove warnings

Uwe-Kohnle committed:
remove warning for scorm

alex40724 committed:
booking manager: display of reservation Information in the tab reservations

alex40724 committed:
feature multi lang wikis

alex40724 committed:
feature feature download selected submissions; feature Submissions and Grades filter; feature always show received peer feedback; panels for exercise view; messages for peer review; indiv. deadline mode; Set grade for multiple assignments

alex40724 committed:
feature feature messages for peer feedback

alex40724 committed:
feature comments in mediacast

alex40724 committed:
feature ::before selector for lists and divs

alex40724 committed:
feature item group: Show applied didactic template for item groups in settings

alex40724 committed:
feature feature messages for peer feedback

alex40724 committed:
empty page fix

alex40724 committed:
feature comments in mediacast

alex40724 committed:
missing lang vars; dtd fix

Uwe-Kohnle committed:
modifications for ui and certificate

alex40724 committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

alex40724 committed:
fixed types

alex40724 committed:
wiki: fixed side block

Uwe-Kohnle committed:
improvements for php 8.1 and 8.2

Uwe-Kohnle committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

Uwe-Kohnle committed:
tiny fixes for LTIconsumer

chfsx committed:
[FIX] UploadLimit audience global_roles

chfsx committed:
[FIX] show upload limit in fileobject

chfsx committed:
[FIX] return type declaration

chfsx committed:
[FIX] typo in doccomment

chfsx committed:
Merge branch 'trunk' into feature/9/customizable-maximum-upload-limit

# Conflicts:
#	Modules/File/classes/class.ilObjFileGUI.php
#	src/UI/Implementation/Component/Input/Field/Renderer.php
#	tests/UI/Component/Input/UploadLimitResolverTest.php

chfsx committed:
[FIX] wring max filesize in error-message

chfsx committed:
[FIX] upload Handler getStatus on bool

chfsx committed:
[FIX] DataSize in different locations

chfsx committed:
[FIX] centralized chunked irss upload handler to aply in file-version tab as well

chfsx committed:
[FIX] FileVersion upload must return new revision number

tfamula committed:
UI: improvements for DateTime/Duration Input

* https://github.com/ILIAS-eLearning/ILIAS/pull/6445

mbecker-databay committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

mbecker-databay committed:
LUI: LinkButtons

alex40724 committed:
updated readme/roadmap

alex40724 committed:
updated roadmap/readme

alex40724 committed:
updated roadmap/readme

chfsx committed:
[FIX] expired limit policies

schmitz-ilias committed:
MetaData: add missing property (38272)

alex40724 committed:
fixed types/declarations

alex40724 committed:
fixed types/declarations

alex40724 committed:
fixed declarations; removed old/wrong if statement

alex40724 committed:
updated readme

alex40724 committed:
updated readme

chfsx committed:
Merge branch 'feature/9/customizable-maximum-upload-limit' into trunk

alex40724 committed:
grp: fixed admin mode issue

schmitz-ilias committed:
fix poll not showing up

matthiaskunkel committed:
Added missing lang var and made some gendermainstreaming in German lang file

alex40724 committed:
fixed 38103: Deleting a Comment throws error

alex40724 committed:
removed inline style, used l-bar__container instead

alex40724 committed:
removed inline style, used l-bar__container instead

alex40724 committed:
added standard links as main action; added tests; fixed existing tests

alex40724 committed:
added standard links as main action; added tests; fixed existing tests

alex40724 commented:
@klees @thibsy All things mentioned by @Amstutz are implemented now. I added Standard Links as actions, too. The testrail case https://testrail.ilias.de//index.php?/cases/view/31882 has been updated accordingly. From my side this is ready for a final review.

alex40724 committed:
removed unused line

alex40724 committed:
removed unused line

alex40724 committed:
removed unused line

alex40724 committed:
removed unused line

kergomard committed:
Object: Fix wrong Return Type for Form-Input

See: https://mantis.ilias.de/view.php?id=37845

kergomard committed:
Test: Fix Get Best Solution in Cloze Question

See: https://mantis.ilias.de/view.php?id=38289

kergomard committed:
Test: Fix Call to Non-Existent Function

chfsx committed:
[FIX] 0038302: Collections return resources multiple times

alex40724 committed:
fixed 38221: When switching from custom to reused title in the settings of ContainerReferences, the title is not immediately updated

alex40724 committed:
SVY: adjust english language variable for deleting... (#6408)

... all participant data to avoid confusion about the functionality, as
user could think that the old phrasing could lead to thinking that every participant gets deleted instead of their data. (35538)

Mantis Ticket: https://mantis.ilias.de/view.php?id=35538

alex40724 committed:
Navigation: 37409 check if type of item is a known plugin id (#6409)

chfsx committed:
[FIX] draft should prevent other actions

chfsx committed:
WIP js

chfsx committed:
[FIX] default value for new column

chfsx committed:
WOPI JS update

chfsx committed:
[FIX] clear Apps before storing new

chfsx committed:
[FIX] Repository Test missing parameter

chfsx committed:
[FIX] Window resize (must wait for mainbar)

chfsx committed:
[FIX] overlay color

kergomard committed:
Test: Fix Call to Undefined Method

kergomard committed:
[FEATURE] UICore: remove legacy lightbox functions (#6448)

kergomard committed:
[FIX] UI: move submit button of roundtrip to right side (#6454)

See https://mantis.ilias.de/view.php?id=38322

kergomard committed:
Test: Fix Call to Undefined Method

kergomard committed:
Test: Remove PDFPurpose From Modules.xml

schmitz-ilias committed:
revert useless change

alex40724 committed:
38297: Class '' was not found in the control structure, try Generating optimized autoload files

Building Static Artifacts for ILIAS
===================================

Deprecated: Return type of ilLTIConsumeProviderList::current() should either be compatible with Iterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /Users/alex/htdocs/trunk/ilias/Modules/LTIConsumer/classes/class.ilLTIConsumeProviderList.php on line 476
Build ./src/FileDelivery/artifacts/key_rotation.php...                     [OK]
Build ./src/FileDelivery/artifacts/delivery_method.php...                  [OK]
Build ./Services/Object/artifacts/object_specific_properties.php...        [OK]
Build Services/Component/artifacts/component_data.php...                   [OK]
Build Services/Component/artifacts/plugin_data.php...                      [OK]
ilComponentRepository is initialized and stored into the environment....   [OK]
Build ./Services/EventHandling/artifacts/event_handling_data.php...        [OK]
Build Services/GlobalScreen/artifacts/global_screen_providers.php...       [OK]
Build ./Services/ResourceStorage/artifacts/flavour_data.php...             [OK]
Build ./Services/Style/System/data/data.php...                             [OK]
Build ./Services/UICore/artifacts/ctrl_base_classes.php...                 [OK]
Build ./Services/UICore/artifacts/ctrl_structure.php...                    [OK]
Build ./Services/UICore/artifacts/ctrl_security.php...                     [OK]

 [OK] All static artifacts are build!

Generated optimized autoload files containing 15021 classes to read artifacts.

alex40724 committed:
38324: Call to undefined method ilNoteGUI::getCommentsHTML()

mbecker-databay committed:
LUI: ClozeSplitButton

mbecker-databay committed:
Merge remote-tracking branch 'origin/trunk' into trunk

alex40724 committed:
38292: LTI: Undefined array key 122

thibsy reviewed (changes requested):
Hi @alex40724

Thx for implementing the requested changes.

Please consider the following suggestions. We do understand that suggestions are not common sense currently, but we still kindly ask you to help us keep our shared UI framework in a good shape. Hence, you do not need to follow those, but please indicate shortly why you prefer to do otherwise:

- [ ] `ItemTest` I: Please use a mock for the tests `test_with_main_action_button()` and `test_with_main_action_link()` instead of a real button and link instance.
- [ ] `ItemTest` II: We try to get rid of rendering tests where we compare the entire HTML to check if only a specific piece exists. Please refactor `test_main_action_button()` and `test_main_action_link()` similar to [`WrapperTest::testRenderWrapperWithAdditionalInputs()`](https://github.com/ILIAS-eLearning/ILIAS/blob/9f37d15a6736d66196bfc39e94e81c52958e1d1f/tests/UI/Component/Dropzone/File/WrapperTest.php#L69), which uses the canonical name of a mocked component to search for a hash with `str_contains()` to check if a specific HTML part will be rendered.

Please implement the following changes:

- [ ] `ItemTest` III: #6376 got merged in the meantime, which changed snake-case function names to camel-case. Please do the same for the newly added tests.
- [ ] Conflicts: please rebase onto the latest trunk to resolve the conflicts.

Kind regards,
@klees & @thibsy

schmitz-ilias committed:
AdvMD: type fix (38304)

chfsx committed:
[FIX] wait for wopi to save document

Uwe-Kohnle committed:
lang var changes for xapi 1

chfsx committed:
[FIX] non needed parameter in instance creation

mbecker-databay committed:
Fixing tests

mbecker-databay committed:
Merge remote-tracking branch 'origin/trunk' into trunk

schmitz-ilias committed:
fix dynamic properties (38240)

klees committed:
Data: add Range::croppedTo

Amstutz committed:
[FIX] UI: select input only displays null-option if necessary (#6450)

See https://mantis.ilias.de/view.php?id=38310

klees committed:
Folders: add 'wopi' as legitimate root-folder

chfsx committed:
[FIX] disabled test due to impossible static mocking of ilLink

alex40724 committed:
mocks; case;

alex40724 committed:
mocks; case;

alex40724 committed:
rebased branch

alex40724 commented:
Rebase went wrong.... Please check https://github.com/ILIAS-eLearning/ILIAS/pull/6460 instead.
